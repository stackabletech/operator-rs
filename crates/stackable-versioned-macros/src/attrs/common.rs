use darling::{
    util::{Flag, Override, SpannedValue},
    Error, FromMeta, Result,
};
use itertools::Itertools;
use k8s_version::Version;

#[derive(Debug, FromMeta)]
#[darling(and_then = CommonRootArguments::validate)]
pub(crate) struct CommonRootArguments {
    #[darling(default)]
    pub(crate) options: RootOptions,

    #[darling(multiple, rename = "version")]
    pub(crate) versions: SpannedValue<Vec<VersionArguments>>,
}

impl CommonRootArguments {
    fn validate(mut self) -> Result<Self> {
        let mut errors = Error::accumulator();

        if self.versions.is_empty() {
            errors.push(
                Error::custom("at least one or more `version`s must be defined")
                    .with_span(&self.versions.span()),
            );
        }

        let is_sorted = self.versions.iter().is_sorted_by_key(|v| v.name);

        // It needs to be sorted, even tho the definition could be unsorted (if allow_unsorted is
        // set).
        self.versions.sort_by(|lhs, rhs| lhs.name.cmp(&rhs.name));

        if !self.options.allow_unsorted.is_present() && !is_sorted {
            let versions = self.versions.iter().map(|v| v.name).join(", ");

            errors.push(Error::custom(format!(
                "versions must be defined in ascending order: {versions}",
            )));
        }

        let duplicate_versions: Vec<_> = self
            .versions
            .iter()
            .duplicates_by(|v| v.name)
            .map(|v| v.name)
            .collect();

        if !duplicate_versions.is_empty() {
            let versions = duplicate_versions.iter().join(", ");

            errors.push(Error::custom(format!(
                "contains duplicate versions: {versions}",
            )));
        }

        errors.finish_with(self)
    }
}

#[derive(Clone, Debug, Default, FromMeta)]
pub(crate) struct RootOptions {
    pub(crate) allow_unsorted: Flag,
    pub(crate) skip: Option<SkipArguments>,
}

/// This struct contains supported version arguments.
///
/// Supported arguments are:
///
/// - `name` of the version, like `v1alpha1`.
/// - `deprecated` flag to mark that version as deprecated.
/// - `skip` option to skip generating various pieces of code.
/// - `doc` option to add version-specific documentation.
#[derive(Clone, Debug, FromMeta)]
pub(crate) struct VersionArguments {
    pub(crate) deprecated: Option<Override<String>>,
    pub(crate) name: Version,
    pub(crate) skip: Option<SkipArguments>,
    pub(crate) doc: Option<String>,
}

/// This struct contains supported common skip arguments.
///
/// Supported arguments are:
///
/// - `from` flag, which skips generating [`From`] implementations when provided.
#[derive(Clone, Debug, Default, FromMeta)]
pub(crate) struct SkipArguments {
    /// Whether the [`From`] implementation generation should be skipped for all versions of this
    /// container.
    pub(crate) from: Flag,
}
