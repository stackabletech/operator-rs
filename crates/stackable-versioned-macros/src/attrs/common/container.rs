use std::{cmp::Ordering, ops::Deref};

use darling::{
    util::{Flag, SpannedValue},
    Error, FromMeta, Result,
};
use itertools::Itertools;
use k8s_version::Version;

/// This struct contains supported container attributes.
///
/// Currently supported attributes are:
///
/// - `version`, which can occur one or more times. See [`VersionAttributes`].
/// - `options`, which allow further customization of the generated code. See [`ContainerOptions`].
#[derive(Debug, FromMeta)]
#[darling(and_then = ContainerAttributes::validate)]
pub(crate) struct ContainerAttributes {
    #[darling(multiple, rename = "version")]
    pub(crate) versions: SpannedValue<Vec<VersionAttributes>>,

    #[darling(default)]
    pub(crate) options: ContainerOptions,
}

impl ContainerAttributes {
    fn validate(mut self) -> Result<Self> {
        // Most of the validation for individual version strings is done by the
        // k8s-version crate. That's why the code below only checks that at
        // least one version is defined, they are defined in order (to ensure
        // code consistency) and that all declared versions are unique.

        // If there are no versions defined, the derive macro errors out. There
        // should be at least one version if the derive macro is used.
        if self.versions.is_empty() {
            return Err(Error::custom(
                "attribute macro `#[versioned()]` must contain at least one `version`",
            )
            .with_span(&self.versions.span()));
        }

        // NOTE (@Techassi): Do we even want to allow to opt-out of this?

        // Ensure that versions are defined in sorted (ascending) order to keep
        // code consistent.
        if !self.options.allow_unsorted.is_present() {
            let original = self.versions.deref().clone();
            self.versions
                .sort_by(|lhs, rhs| lhs.name.partial_cmp(&rhs.name).unwrap_or(Ordering::Equal));

            for (index, version) in original.iter().enumerate() {
                if version.name
                    == self
                        .versions
                        .get(index)
                        .expect("internal error: version at that index must exist")
                        .name
                {
                    continue;
                }

                return Err(Error::custom(format!(
                    "versions in `#[versioned()]` must be defined in ascending order (version `{name}` is misplaced)",
                    name = version.name
                )));
            }
        }

        // TODO (@Techassi): Add validation for skip(from) for last version,
        // which will skip nothing, because nothing is generated in the first
        // place.

        // Ensure every version is unique and isn't declared multiple times.
        let duplicates = self
            .versions
            .iter()
            .duplicates_by(|e| e.name)
            .map(|e| e.name)
            .join(", ");

        if !duplicates.is_empty() {
            return Err(Error::custom(format!(
                "attribute macro `#[versioned()]` contains duplicate versions: {duplicates}",
            ))
            .with_span(&self.versions.span()));
        }

        Ok(self)
    }
}

/// This struct contains supported version options.
///
/// Supported options are:
///
/// - `name` of the version, like `v1alpha1`.
/// - `deprecated` flag to mark that version as deprecated.
/// - `skip` option to skip generating various pieces of code.
/// - `doc` option to add version-specific documentation.
#[derive(Clone, Debug, FromMeta)]
pub(crate) struct VersionAttributes {
    pub(crate) deprecated: Flag,
    pub(crate) name: Version,
    pub(crate) skip: Option<SkipOptions>,
    pub(crate) doc: Option<String>,
}

/// This struct contains supported container options.
///
/// Supported options are:
///
/// - `allow_unsorted`, which allows declaring versions in unsorted order,
///   instead of enforcing ascending order.
/// - `skip` option to skip generating various pieces of code.
#[derive(Clone, Debug, Default, FromMeta)]
pub(crate) struct ContainerOptions {
    pub(crate) allow_unsorted: Flag,
    pub(crate) skip: Option<SkipOptions>,
}

/// This struct contains supported skip options.
///
/// Supported options are:
///
/// - `from` flag, which skips generating [`From`] implementations when provided.
#[derive(Clone, Debug, Default, FromMeta)]
pub(crate) struct SkipOptions {
    pub(crate) from: Flag,
}
