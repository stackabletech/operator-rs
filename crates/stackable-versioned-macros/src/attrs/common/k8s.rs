use darling::{util::Flag, FromMeta};

/// This struct contains supported Kubernetes arguments.
///
/// Supported arguments are:
///
/// - `skip`, which controls skipping parts of the generation.
/// - `kind`, which allows overwriting the kind field of the CRD. This defaults to the struct name
///    (without the 'Spec' suffix).
/// - `group`, which sets the CRD group, usually the domain of the company.
#[derive(Clone, Debug, FromMeta)]
pub(crate) struct KubernetesArguments {
    pub(crate) skip: Option<KubernetesSkipArguments>,
    pub(crate) singular: Option<String>,
    pub(crate) plural: Option<String>,
    pub(crate) kind: Option<String>,
    pub(crate) namespaced: Flag,
    pub(crate) group: String,
}

/// This struct contains supported kubernetes skip arguments.
///
/// Supported arguments are:
///
/// - `merged_crd` flag, which skips generating the `crd()` and `merged_crd()` functions are
///    generated.
#[derive(Clone, Debug, FromMeta)]
pub(crate) struct KubernetesSkipArguments {
    /// Whether the `crd()` and `merged_crd()` generation should be skipped for
    /// this container.
    pub(crate) merged_crd: Flag,
}
