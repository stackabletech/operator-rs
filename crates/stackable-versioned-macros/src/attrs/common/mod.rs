mod container;
mod item;

pub(crate) use container::*;
pub(crate) use item::*;
