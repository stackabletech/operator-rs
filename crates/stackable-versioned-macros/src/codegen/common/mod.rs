use std::collections::BTreeMap;

use k8s_version::Version;
use proc_macro2::Span;
use quote::format_ident;
use syn::Ident;

use crate::{
    attrs::common::ContainerAttributes,
    consts::{DEPRECATED_FIELD_PREFIX, DEPRECATED_VARIANT_PREFIX},
};

mod container;
mod item;

pub(crate) use container::*;
pub(crate) use item::*;

/// Type alias to make the type of the version chain easier to handle.
pub(crate) type VersionChain = BTreeMap<Version, ItemStatus>;

#[derive(Debug, Clone)]
pub(crate) struct ContainerVersion {
    /// Indicates that the container version is deprecated.
    pub(crate) deprecated: bool,

    /// Indicates that the generation of `From<OLD> for NEW` should be skipped.
    pub(crate) skip_from: bool,

    /// A validated Kubernetes API version.
    pub(crate) inner: Version,

    /// The ident of the container.
    pub(crate) ident: Ident,
}

impl From<&ContainerAttributes> for Vec<ContainerVersion> {
    fn from(attributes: &ContainerAttributes) -> Self {
        attributes
            .versions
            .iter()
            .map(|v| ContainerVersion {
                skip_from: v.skip.as_ref().map_or(false, |s| s.from.is_present()),
                ident: Ident::new(&v.name.to_string(), Span::call_site()),
                deprecated: v.deprecated.is_present(),
                inner: v.name,
            })
            .collect()
    }
}

/// Returns the container ident used in [`From`] implementations.
pub(crate) fn format_container_from_ident(ident: &Ident) -> Ident {
    format_ident!("__sv_{ident}", ident = ident.to_string().to_lowercase())
}

/// Removes the deprecated prefix from a field ident.
///
/// See [`DEPRECATED_FIELD_PREFIX`].
pub(crate) fn remove_deprecated_field_prefix(ident: &Ident) -> Ident {
    remove_ident_prefix(ident, DEPRECATED_FIELD_PREFIX)
}

/// Removes the deprecated prefix from a variant ident.
///
/// See [`DEPRECATED_VARIANT_PREFIX`].
pub(crate) fn remove_deprecated_variant_prefix(ident: &Ident) -> Ident {
    remove_ident_prefix(ident, DEPRECATED_VARIANT_PREFIX)
}

/// Removes the provided prefix from an ident and returns the newly created
/// ident.
pub(crate) fn remove_ident_prefix(ident: &Ident, prefix: &str) -> Ident {
    format_ident!("{}", ident.to_string().trim_start_matches(prefix))
}
