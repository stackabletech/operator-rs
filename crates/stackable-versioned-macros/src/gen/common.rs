use std::{collections::BTreeMap, ops::Deref};

use k8s_version::Version;
use proc_macro2::{Span, TokenStream};
use quote::format_ident;
use syn::{Ident, Path};

use crate::{
    attrs::common::ContainerAttributes,
    consts::{DEPRECATED_FIELD_PREFIX, DEPRECATED_VARIANT_PREFIX},
};

pub(crate) type VersionChain = BTreeMap<Version, ItemStatus>;

#[derive(Debug, Clone)]
pub(crate) struct ContainerVersion {
    pub(crate) deprecated: bool,
    pub(crate) skip_from: bool,
    pub(crate) inner: Version,
    pub(crate) ident: Ident,
}

impl From<&ContainerAttributes> for Vec<ContainerVersion> {
    fn from(attributes: &ContainerAttributes) -> Self {
        attributes
            .versions
            .iter()
            .map(|v| ContainerVersion {
                skip_from: v.skip.as_ref().map_or(false, |s| s.from.is_present()),
                ident: Ident::new(&v.name.to_string(), Span::call_site()),
                deprecated: v.deprecated.is_present(),
                inner: v.name,
            })
            .collect()
    }
}

pub(crate) trait Container<D, I>
where
    Self: Sized + Deref<Target = VersionedContainer<I>>,
{
    fn new(ident: Ident, data: D, attributes: ContainerAttributes) -> syn::Result<Self>;

    /// This generates the complete code for a single versioned container.
    ///
    /// Internally, it will create a module for each declared version which
    /// contains the container with the appropriate items (fields or variants)
    /// Additionally, it generates `From` implementations, which enable
    /// conversion from an older to a newer version.
    fn generate_tokens(&self) -> TokenStream;
}

#[derive(Debug)]
pub(crate) struct VersionedContainer<I> {
    pub(crate) versions: Vec<ContainerVersion>,
    pub(crate) items: Vec<I>,
    pub(crate) ident: Ident,

    pub(crate) from_ident: Ident,
    pub(crate) skip_from: bool,
}

#[derive(Debug)]
pub(crate) enum ItemStatus {
    Added {
        ident: Ident,
        default_fn: Path,
    },
    Renamed {
        from: Ident,
        to: Ident,
    },
    Deprecated {
        previous_ident: Ident,
        ident: Ident,
        note: String,
    },
    NoChange(Ident),
    NotPresent,
}

impl ItemStatus {
    pub(crate) fn get_ident(&self) -> Option<&Ident> {
        match &self {
            ItemStatus::Added { ident, .. } => Some(ident),
            ItemStatus::Renamed { to, .. } => Some(to),
            ItemStatus::Deprecated { ident, .. } => Some(ident),
            ItemStatus::NoChange(ident) => Some(ident),
            ItemStatus::NotPresent => None,
        }
    }
}

/// Returns the container ident used in [`From`] implementations.
pub(crate) fn format_container_from_ident(ident: &Ident) -> Ident {
    format_ident!("__sv_{ident}", ident = ident.to_string().to_lowercase())
}

/// Removes the deprecated prefix from field ident.
pub(crate) fn remove_deprecated_field_prefix(ident: &Ident) -> Ident {
    remove_ident_prefix(ident, DEPRECATED_FIELD_PREFIX)
}

pub(crate) fn remove_deprecated_variant_prefix(ident: &Ident) -> Ident {
    remove_ident_prefix(ident, DEPRECATED_VARIANT_PREFIX)
}

pub(crate) fn remove_ident_prefix(ident: &Ident, prefix: &str) -> Ident {
    format_ident!("{}", ident.to_string().trim_start_matches(prefix))
}
