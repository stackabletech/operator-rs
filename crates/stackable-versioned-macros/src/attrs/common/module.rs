use darling::{
    util::{Flag, SpannedValue},
    FromMeta, Result,
};

use crate::attrs::common::{SkipArguments, VersionArguments};

#[derive(Debug, FromMeta)]
#[darling(and_then = ModuleAttributes::validate)]
pub(crate) struct ModuleAttributes {
    #[darling(multiple, rename = "version")]
    pub(crate) versions: SpannedValue<Vec<VersionArguments>>,

    #[darling(default, rename = "options")]
    pub(crate) common_option_args: ModuleOptionArguments,
}

impl ModuleAttributes {
    fn validate(self) -> Result<Self> {
        // TODO (@Techassi): Make this actually validate
        Ok(self)
    }
}

#[derive(Debug, Default, FromMeta)]
pub(crate) struct ModuleOptionArguments {
    pub(crate) skip: Option<SkipArguments>,
    pub(crate) preserve_module: Flag,
    pub(crate) allow_unsorted: Flag,
}
