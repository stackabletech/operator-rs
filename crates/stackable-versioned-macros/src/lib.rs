use darling::{ast::NestedMeta, FromMeta};
use proc_macro::TokenStream;
use syn::{DeriveInput, Error};

use crate::attrs::common::ContainerAttributes;

mod attrs;
mod consts;
mod gen;

/// This macro enables generating versioned structs.
///
/// ## Usage Guide
///
/// ### Quickstart
///
/// ```
/// # use stackable_versioned_macros::versioned;
/// #[versioned(
///     version(name = "v1alpha1"),
///     version(name = "v1beta1"),
///     version(name = "v1"),
///     version(name = "v2"),
///     version(name = "v3")
/// )]
/// struct Foo {
///     /// My docs
///     #[versioned(
///         added(since = "v1beta1"),
///         renamed(since = "v1", from = "gau"),
///         deprecated(since = "v2", note = "not empty")
///     )]
///     deprecated_bar: usize,
///     baz: bool,
/// }
/// ```
///
/// ### Declaring Versions
///
/// Before any of the fields can be versioned, versions need to be declared at
/// the container level. Each version currently supports two parameters: `name`
/// and the `deprecated` flag. The `name` must be a valid (and supported)
/// format. The macro checks each declared version and reports any error
/// encountered during parsing.
/// The `deprecated` flag marks the version as deprecated. This currently adds
/// the `#[deprecated]` attribute to the appropriate piece of code.
///
/// ```
/// # use stackable_versioned_macros::versioned;
/// #[versioned(
///     version(name = "v1alpha1", deprecated)
/// )]
/// struct Foo {}
/// ```
///
/// Additionally, it is ensured that each version is unique. Declaring the same
/// version multiple times will result in an error. Furthermore, declaring the
/// versions out-of-order is prohibited by default. It is possible to opt-out
/// of this check by setting `options(allow_unsorted)`:
///
/// ```
/// # use stackable_versioned_macros::versioned;
/// #[versioned(
///     version(name = "v1beta1"),
///     version(name = "v1alpha1"),
///     options(allow_unsorted)
/// )]
/// struct Foo {}
/// ```
///
/// ### Field Actions
///
/// This library currently supports three different field actions. Fields can
/// be added, renamed and deprecated. The macro ensures that these actions
/// adhere to the following set of rules:
///
/// - Fields cannot be added and deprecated in the same version.
/// - Fields cannot be added and renamed in the same version.
/// - Fields cannot be renamed and deprecated in the same version.
/// - Fields added in version _a_, renamed _0...n_ times in versions
///   b<sub>1</sub>, b<sub>2</sub>, ..., b<sub>n</sub> and deprecated in
///   version _c_ must ensure _a < b<sub>1</sub>, b<sub>2</sub>, ...,
///   b<sub>n</sub> < c_.
/// - All field actions must use previously declared versions. Using versions
///   not present at the container level will result in an error.
///
/// For fields marked as deprecated, two additional rules apply:
///
/// - Fields must start with the `deprecated_` prefix.
/// - The deprecation note cannot be empty.
///
/// ### Auto-generated [`From`] Implementations
///
/// To enable smooth version upgrades of the same struct, the macro automatically
/// generates [`From`] implementations. On a high level, code generated for two
/// versions _a_ and _b_, with _a < b_ looks like this: `impl From<a> for b`.
///
/// ```ignore
/// #[versioned(
///     version(name = "v1alpha1"),
///     version(name = "v1beta1"),
///     version(name = "v1")
/// )]
/// pub struct Foo {
///     #[versioned(
///         added(since = "v1beta1"),
///         deprecated(since = "v1", note = "not needed")
///     )]
///     deprecated_bar: usize,
///     baz: bool,
/// }
///
/// // Produces ...
///
/// #[automatically_derived]
/// pub mod v1alpha1 {
///     pub struct Foo {
///         pub baz: bool,
///     }
/// }
/// #[automatically_derived]
/// #[allow(deprecated)]
/// impl From<v1alpha1::Foo> for v1beta1::Foo {
///     fn from(__sv_foo: v1alpha1::Foo) -> Self {
///         Self {
///             bar: std::default::Default::default(),
///             baz: __sv_foo.baz,
///         }
///     }
/// }
/// #[automatically_derived]
/// pub mod v1beta1 {
///     pub struct Foo {
///         pub bar: usize,
///         pub baz: bool,
///     }
/// }
/// #[automatically_derived]
/// #[allow(deprecated)]
/// impl From<v1beta1::Foo> for v1::Foo {
///     fn from(__sv_foo: v1beta1::Foo) -> Self {
///         Self {
///             deprecated_bar: __sv_foo.bar,
///             baz: __sv_foo.baz,
///         }
///     }
/// }
/// #[automatically_derived]
/// pub mod v1 {
///     pub struct Foo {
///         #[deprecated = "not needed"]
///         pub deprecated_bar: usize,
///         pub baz: bool,
///     }
/// }
/// ```
///
/// #### Skip [`From`] generation
///
/// Generation of these [`From`] implementations can be skipped at the container
/// and version level. This enables customization of the implementations if the
/// default implementation is not sufficient.
///
/// ```
/// # use stackable_versioned_macros::versioned;
/// #[versioned(
///     version(name = "v1alpha1"),
///     version(name = "v1beta1"),
///     version(name = "v1"),
///     options(skip(from))
/// )]
/// pub struct Foo {
///     #[versioned(
///         added(since = "v1beta1"),
///         deprecated(since = "v1", note = "not needed")
///     )]
///     deprecated_bar: usize,
///     baz: bool,
/// }
/// ```
///
/// #### Customize Default Function for Added Fields
///
/// It is possible to customize the default function used in the generated
/// [`From`] implementation for populating added fields. By default,
/// [`Default::default()`] is used.
///
/// ```
/// # use stackable_versioned_macros::versioned;
/// #[versioned(
///     version(name = "v1alpha1"),
///     version(name = "v1beta1"),
///     version(name = "v1")
/// )]
/// pub struct Foo {
///     #[versioned(
///         added(since = "v1beta1", default = "default_bar"),
///         deprecated(since = "v1", note = "not needed")
///     )]
///     deprecated_bar: usize,
///     baz: bool,
/// }
///
/// fn default_bar() -> usize {
///     42
/// }
/// ```
#[proc_macro_attribute]
pub fn versioned(attrs: TokenStream, input: TokenStream) -> TokenStream {
    let attrs = match NestedMeta::parse_meta_list(attrs.into()) {
        Ok(attrs) => match ContainerAttributes::from_list(&attrs) {
            Ok(attrs) => attrs,
            Err(err) => return err.write_errors().into(),
        },
        Err(err) => return darling::Error::from(err).write_errors().into(),
    };

    // NOTE (@Techassi): For now, we can just use the DeriveInput type here,
    // because we only support structs (and eventually enums) to be versioned.
    // In the future - if we decide to support modules - this requires
    // adjustments to also support modules. One possible solution might be to
    // use an enum with two variants: Container(DeriveInput) and
    // Module(ItemMod).
    let input = syn::parse_macro_input!(input as DeriveInput);

    gen::expand(attrs, input)
        .unwrap_or_else(Error::into_compile_error)
        .into()
}
