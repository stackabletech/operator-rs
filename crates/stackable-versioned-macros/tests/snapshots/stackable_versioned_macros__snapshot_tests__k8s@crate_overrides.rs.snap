---
source: crates/stackable-versioned-macros/src/lib.rs
expression: formatted
input_file: crates/stackable-versioned-macros/tests/inputs/k8s/pass/crate_overrides.rs
---
#[automatically_derived]
pub mod v1alpha1 {
    use super::*;
    #[derive(
        Clone,
        Debug,
        serde::Deserialize,
        serde::Serialize,
        schemars::JsonSchema,
        kube::CustomResource,
    )]
    #[kube(
        group = "foo.example.org",
        version = "v1alpha1",
        kind = "Foo",
        crates(kube_core = ::kube::core, schemars = ::schemars)
    )]
    pub struct FooSpec {
        pub baz: bool,
    }
}
#[automatically_derived]
impl ::std::convert::From<v1alpha1::FooSpec> for v1beta1::FooSpec {
    fn from(__sv_foospec: v1alpha1::FooSpec) -> Self {
        Self {
            bah: ::std::default::Default::default(),
            baz: __sv_foospec.baz.into(),
        }
    }
}
#[automatically_derived]
impl ::std::convert::From<v1beta1::FooSpec> for v1alpha1::FooSpec {
    fn from(__sv_foospec: v1beta1::FooSpec) -> Self {
        Self {
            baz: __sv_foospec.baz.into(),
        }
    }
}
#[automatically_derived]
pub mod v1beta1 {
    use super::*;
    #[derive(
        Clone,
        Debug,
        serde::Deserialize,
        serde::Serialize,
        schemars::JsonSchema,
        kube::CustomResource,
    )]
    #[kube(
        group = "foo.example.org",
        version = "v1beta1",
        kind = "Foo",
        crates(kube_core = ::kube::core, schemars = ::schemars)
    )]
    pub struct FooSpec {
        pub bah: usize,
        pub baz: bool,
    }
}
#[automatically_derived]
impl ::std::convert::From<v1beta1::FooSpec> for v1::FooSpec {
    fn from(__sv_foospec: v1beta1::FooSpec) -> Self {
        Self {
            bar: __sv_foospec.bah.into(),
            baz: __sv_foospec.baz.into(),
        }
    }
}
#[automatically_derived]
impl ::std::convert::From<v1::FooSpec> for v1beta1::FooSpec {
    fn from(__sv_foospec: v1::FooSpec) -> Self {
        Self {
            bah: __sv_foospec.bar.into(),
            baz: __sv_foospec.baz.into(),
        }
    }
}
#[automatically_derived]
pub mod v1 {
    use super::*;
    #[derive(
        Clone,
        Debug,
        serde::Deserialize,
        serde::Serialize,
        schemars::JsonSchema,
        kube::CustomResource,
    )]
    #[kube(
        group = "foo.example.org",
        version = "v1",
        kind = "Foo",
        crates(kube_core = ::kube::core, schemars = ::schemars)
    )]
    pub struct FooSpec {
        pub bar: usize,
        pub baz: bool,
    }
}
#[automatically_derived]
pub enum Foo {
    V1Alpha1(v1alpha1::Foo),
    V1Beta1(v1beta1::Foo),
    V1(v1::Foo),
}
#[automatically_derived]
impl Foo {
    /// Generates a merged CRD containing all versions and marking `stored_apiversion` as stored.
    pub fn merged_crd(
        stored_apiversion: FooVersion,
    ) -> ::std::result::Result<
        ::k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition,
        ::kube::core::crd::MergeError,
    > {
        ::kube::core::crd::merge_crds(
            vec![
                < v1alpha1::Foo as ::kube::core::CustomResourceExt > ::crd(), <
                v1beta1::Foo as ::kube::core::CustomResourceExt > ::crd(), < v1::Foo as
                ::kube::core::CustomResourceExt > ::crd()
            ],
            stored_apiversion.as_str(),
        )
    }
    ///Tries to convert a list of objects of kind [`Foo`] to the desired API version
    ///specified in the [`ConversionReview`][cr].
    ///
    ///The returned [`ConversionReview`][cr] either indicates a success or a failure, which
    ///is handed back to the Kubernetes API server.
    ///
    ///[cr]: ::kube::core::conversion::ConversionReview
    pub fn try_convert(
        review: ::kube::core::conversion::ConversionReview,
    ) -> ::kube::core::conversion::ConversionReview {
        let request = match ::kube::core::conversion::ConversionRequest::from_review(
            review,
        ) {
            ::std::result::Result::Ok(request) => request,
            ::std::result::Result::Err(err) => {
                return ::kube::core::conversion::ConversionResponse::invalid(::kube::client::Status {
                        status: Some(::kube::core::response::StatusSummary::Failure),
                        message: err.to_string(),
                        reason: err.to_string(),
                        details: None,
                        code: 400,
                    })
                    .into_review();
            }
        };
        let desired_api_version = request.desired_api_version.as_str();
        let response = match Self::convert_objects(
            request.objects,
            desired_api_version,
        ) {
            ::std::result::Result::Ok(converted_objects) => {
                ::kube::core::conversion::ConversionResponse {
                    result: ::kube::client::Status::success(),
                    types: request.types,
                    uid: request.uid,
                    converted_objects,
                }
            }
            ::std::result::Result::Err(err) => {
                let code = err.http_status_code();
                let message = err.join_errors();
                ::kube::core::conversion::ConversionResponse {
                    result: ::kube::client::Status {
                        status: Some(::kube::core::response::StatusSummary::Failure),
                        message: message.clone(),
                        reason: message,
                        details: None,
                        code,
                    },
                    types: request.types,
                    uid: request.uid,
                    converted_objects: vec![],
                }
            }
        };
        response.into_review()
    }
    fn convert_objects(
        objects: ::std::vec::Vec<::serde_json::Value>,
        desired_api_version: &str,
    ) -> ::std::result::Result<
        ::std::vec::Vec<::serde_json::Value>,
        ::stackable_versioned::ConvertObjectError,
    > {
        let mut converted_objects = ::std::vec::Vec::with_capacity(objects.len());
        for object in objects {
            let current_object = Self::from_json_value(object.clone())
                .map_err(|source| ::stackable_versioned::ConvertObjectError::Parse {
                    source,
                })?;
            match (current_object, desired_api_version) {
                (Self::V1Alpha1(__sv_foo), "v1beta1") => {
                    let converted: v1beta1::FooSpec = __sv_foo.spec.into();
                    let desired_object = Self::V1Beta1(v1beta1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                (Self::V1Alpha1(__sv_foo), "v1") => {
                    let converted: v1beta1::FooSpec = __sv_foo.spec.into();
                    let converted: v1::FooSpec = converted.into();
                    let desired_object = Self::V1(v1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                (Self::V1Beta1(__sv_foo), "v1alpha1") => {
                    let converted: v1alpha1::FooSpec = __sv_foo.spec.into();
                    let desired_object = Self::V1Alpha1(v1alpha1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                (Self::V1Beta1(__sv_foo), "v1") => {
                    let converted: v1::FooSpec = __sv_foo.spec.into();
                    let desired_object = Self::V1(v1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                (Self::V1(__sv_foo), "v1alpha1") => {
                    let converted: v1beta1::FooSpec = __sv_foo.spec.into();
                    let converted: v1alpha1::FooSpec = converted.into();
                    let desired_object = Self::V1Alpha1(v1alpha1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                (Self::V1(__sv_foo), "v1beta1") => {
                    let converted: v1beta1::FooSpec = __sv_foo.spec.into();
                    let desired_object = Self::V1Beta1(v1beta1::Foo {
                        metadata: __sv_foo.metadata,
                        spec: converted,
                    });
                    let desired_object = desired_object
                        .into_json_value()
                        .map_err(|source| ::stackable_versioned::ConvertObjectError::Serialize {
                            source,
                        })?;
                    converted_objects.push(desired_object);
                }
                _ => converted_objects.push(object),
            }
        }
        ::std::result::Result::Ok(converted_objects)
    }
    fn from_json_value(
        value: ::serde_json::Value,
    ) -> ::std::result::Result<Self, ::stackable_versioned::ParseObjectError> {
        let api_version = value
            .get("apiVersion")
            .ok_or_else(|| ::stackable_versioned::ParseObjectError::FieldNotPresent)?
            .as_str()
            .ok_or_else(|| ::stackable_versioned::ParseObjectError::FieldNotStr)?;
        let object = match api_version {
            "foo.example.org/v1alpha1" | "v1alpha1" => {
                let object = ::serde_json::from_value(value)
                    .map_err(|source| ::stackable_versioned::ParseObjectError::Deserialize {
                        source,
                    })?;
                Self::V1Alpha1(object)
            }
            "foo.example.org/v1beta1" | "v1beta1" => {
                let object = ::serde_json::from_value(value)
                    .map_err(|source| ::stackable_versioned::ParseObjectError::Deserialize {
                        source,
                    })?;
                Self::V1Beta1(object)
            }
            "foo.example.org/v1" | "v1" => {
                let object = ::serde_json::from_value(value)
                    .map_err(|source| ::stackable_versioned::ParseObjectError::Deserialize {
                        source,
                    })?;
                Self::V1(object)
            }
            unknown_api_version => {
                return ::std::result::Result::Err(::stackable_versioned::ParseObjectError::UnknownApiVersion {
                    api_version: unknown_api_version.to_owned(),
                });
            }
        };
        ::std::result::Result::Ok(object)
    }
    fn into_json_value(
        self,
    ) -> ::std::result::Result<::serde_json::Value, ::serde_json::Error> {
        match self {
            Self::V1Alpha1(__sv_foo) => Ok(::serde_json::to_value(__sv_foo)?),
            Self::V1Beta1(__sv_foo) => Ok(::serde_json::to_value(__sv_foo)?),
            Self::V1(__sv_foo) => Ok(::serde_json::to_value(__sv_foo)?),
        }
    }
}
#[automatically_derived]
pub enum FooVersion {
    V1Alpha1,
    V1Beta1,
    V1,
}
#[automatically_derived]
impl ::std::fmt::Display for FooVersion {
    fn fmt(
        &self,
        f: &mut ::std::fmt::Formatter<'_>,
    ) -> ::std::result::Result<(), ::std::fmt::Error> {
        f.write_str(self.as_str())
    }
}
#[automatically_derived]
impl FooVersion {
    pub fn as_str(&self) -> &str {
        match self {
            V1Alpha1 => "v1alpha1",
            V1Beta1 => "v1beta1",
            V1 => "v1",
        }
    }
}
